### PR TITLE
GLT-1250 Migrate ReferenceGenome by name

### PR DESCRIPTION
**Pull request by tonydebat**
Allowed ReferenceGenome to be retrieved by id.
Converted ReferenceGenome to charset utf8.
Project now refers to ReferenceGenome as an object and not an id.
Added project resolver to migration allowing a ReferenceGenome to be specified by name during migration.
Updated edit project UI to work with ReferenceGenome changes. Added a binding to convert by id.

### CODE REVIEW

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/AbstractProject.java [hunk 1]
```diff
@@ -469,7 +473,7 @@ public int hashCode() {
         .append(alias)
         .append(description)
         .append(progress)
-        .append(referenceGenomeId)
+        .append(referenceGenome.getId())
```
Comment (djcooke) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/AbstractProject.java]: Potential NPE. Also won't be useful when the ID is not set (such as during migration). Could just append the ReferenceGenome object here, and then base ReferenceGenome's hashCode on the alias (which is a unique key also)

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/AbstractProject.java [hunk 2]
```diff
@@ -484,7 +488,7 @@ public boolean equals(Object obj) {
         .append(alias, other.alias)
         .append(description, other.description)
         .append(progress, other.progress)
-        .append(referenceGenomeId, other.referenceGenomeId)
+        .append(referenceGenome.getId(), other.referenceGenome.getId())
```
Comment (djcooke) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/AbstractProject.java]: Same issue as hashcode

File: migration/src/main/java/uk/ac/bbsrc/tgac/miso/migration/destination/MisoServiceManager.java [hunk 3]
```diff
@@ -1261,6 +1271,40 @@ private void updateSampleValidRelationshipServiceDependencies() {
...
+  public HibernateReferenceGenomeDao getReferenceGenomeDao() {
+    return referenceGenomeDao;
+  }
+
+  public void setReferenceGenomeDao(HibernateReferenceGenomeDao referenceGenomeDao) {
+    this.referenceGenomeDao = referenceGenomeDao;
+    updateReferenceGenomeDaoDependencies();
+  }
+
+  public void setReferenceGenomeDao() {
```
Comment (djcooke) [inline migration/src/main/java/uk/ac/bbsrc/tgac/miso/migration/destination/MisoServiceManager.java]: `setDefaultReferenceGenomeDao` to match other methods' naming

File: migration/src/main/java/uk/ac/bbsrc/tgac/miso/migration/destination/MisoServiceManager.java [hunk 4]
```diff
@@ -414,12 +422,14 @@ public void setDefaultProjectDao() {
...
     dao.setWatcherDAO(watcherDao);
     dao.setDataObjectFactory(dataObjectFactory);
+    dao.setReferenceGenomeDao(referenceGenomeDao);
     setProjectDao(dao);
   }
 
   private void updateProjectDaoDependencies() {
     if (studyDao != null) studyDao.setProjectDAO(projectDao);
     if (sampleNumberPerProjectService != null) sampleNumberPerProjectService.setSqlProjectDAO(projectDao);
+    if (referenceGenomeService != null) referenceGenomeService.setReferenceGenomeDao(referenceGenomeDao);
```
Comment (djcooke) [inline migration/src/main/java/uk/ac/bbsrc/tgac/miso/migration/destination/MisoServiceManager.java]: This method is for updating other objects' dependencies on the Project DAO, so this doesn't belong here. There should be an `updateReferenceGenomeDaoDependencies` method which sets the referenceGenomeDao for projectDao and referenceGenomeService

File: migration/src/main/java/uk/ac/bbsrc/tgac/miso/migration/destination/ValueTypeLookup.java [hunk 5]
```diff
@@ -900,4 +914,16 @@ public void resolveAll(Run run) throws IOException {
     }
   }
 
+  public void resolveAll(Project project) {
+    ReferenceGenome referenceGenome = resolve(project.getReferenceGenome());
+    project.setReferenceGenome(referenceGenome);
+  }
+
+  @VisibleForTesting
+  ReferenceGenome resolve(ReferenceGenome referenceGenome) {
```
Comment (djcooke) [inline migration/src/main/java/uk/ac/bbsrc/tgac/miso/migration/destination/ValueTypeLookup.java]: should be public

Comment (djcooke) [inline migration/src/main/java/uk/ac/bbsrc/tgac/miso/migration/destination/ValueTypeLookup.java]: Add Javadoc as well. Lookup is only possible by alias, whereas most types can be found by ID as well

File: migration/src/main/java/uk/ac/bbsrc/tgac/miso/migration/source/LoadGeneratorSource.java [hunk 6]
```diff
@@ -499,6 +501,13 @@ private Run createRun(List<Pool> pools, int runNum) {
     return run;
   }
 
+  private ReferenceGenome createReferenceGenome() {
+    ReferenceGenome referenceGenome = new ReferenceGenomeImpl();
+    referenceGenome.setAlias("Human hg19");
+    referenceGenome.setId(1L);
```
Comment (djcooke) [inline migration/src/main/java/uk/ac/bbsrc/tgac/miso/migration/source/LoadGeneratorSource.java]: Shouldn't need to specify ID anymore

**Comment by tonydebat** [comment]
Dillan, I've made the changes you requested.

File: migration/src/main/java/uk/ac/bbsrc/tgac/miso/migration/destination/MisoServiceManager.java [hunk 7]
```diff
@@ -1261,6 +1270,40 @@ private void updateSampleValidRelationshipServiceDependencies() {
...
+    updateReferenceGenomeDaoDependencies();
+  }
+
+  public void setDefaultReferenceGenomeDao() {
+    HibernateReferenceGenomeDao dao = new HibernateReferenceGenomeDao();
+    dao.setSessionFactory(sessionFactory);
+    setReferenceGenomeDao(dao);
+  }
+
+  private void updateReferenceGenomeDaoDependencies() {
```
Comment (djcooke) [inline migration/src/main/java/uk/ac/bbsrc/tgac/miso/migration/destination/MisoServiceManager.java]: this method should call projectDao.setReferenceGenomeDao as well

**Pull request merged**